panoptes_aggregation/tests/batch_aggregation/test_batch_aggregation.py

```diff
@@ -0,0 +1,124 @@
+import unittest
+from unittest.mock import patch, MagicMock, call
+from panoptes_aggregation.scripts import batch_utils
+from panoptes_aggregation.batch_aggregation import run_aggregation
+from panoptes_aggregation import batch_aggregation as batch_agg
+
+wf_export = 'panoptes_aggregation/tests/batch_aggregation/wf_export.csv'
+cls_export = 'panoptes_aggregation/tests/batch_aggregation/cls_export.csv'
+
+
+@patch("panoptes_aggregation.batch_aggregation.BatchAggregator._connect_api_client", new=MagicMock())
+class TestBatchAggregation(unittest.TestCase):
+    @patch("panoptes_aggregation.batch_aggregation.workflow_extractor_config")
+    @patch("panoptes_aggregation.batch_aggregation.BatchAggregator")
+    def test_run_aggregation(self, mock_aggregator, mock_wf_ext_conf):
+        mock_aggregator.process_wf_export.return_value = MagicMock()
+        mock_aggregator.process_cls_export.return_value = MagicMock()
+
+        mock_df = MagicMock()
+        test_extracts = {'question_extractor': mock_df}
+        batch_utils.batch_extract = MagicMock(return_value=test_extracts)
+        mock_reducer = MagicMock()
+        batch_utils.batch_reduce = mock_reducer
+
+        run_aggregation(1, 10, 100)
+        mock_aggregator.assert_called_once_with(1, 10, 100)
+        mock_wf_ext_conf.assert_called_once()
+        batch_utils.batch_extract.assert_called_once()
+        mock_df.to_csv.assert_called()
+        batch_utils.batch_reduce.assert_called()
+        self.assertEqual(mock_reducer.call_count, 2)
+
+        # The reducer's call list includes subsequent calls to to_csv, but the args are methods called on the mock
+        # rather than use the set values i.e. "<MagicMock name='BatchAggregator().output_path' id='140281634764400'>"
```

Comment:
You might be able to put a mock directly on the `to_csv` function from the pandas to see if that allows you to intercept the calls.

From https://stackoverflow.com/a/65593458/1052418
>You could mock out the entire DataFrame class using mock.patch("pandas.DataFrame", ...).

or more specifically `mock.patch("pandas.DataFrame.to_csv", ...)`

Reply:
Also if mocking to far up is making some tests difficult to write you can split this into multiple tests that place the mocks at different levels for checking different things.

Reply:
Mostly this was confusion about why the to_csv was in the same call stack but as a different mock. As with the other issue, seems like it was just a different mock of the same thing. With all the other spec changes, I just removed this block and I'm satisfied if you are.